common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java

```diff
@@ -138,9 +138,24 @@ private FetchShuffleBlocks createFetchShuffleBlocksMsg(
     }
     long[] mapIds = Longs.toArray(mapIdToReduceIds.keySet());
     int[][] reduceIdArr = new int[mapIds.length][];
+    int blockIdIndex = 0;
     for (int i = 0; i < mapIds.length; i++) {
       reduceIdArr[i] = Ints.toArray(mapIdToReduceIds.get(mapIds[i]));
+      // The `blockIds`'s order must be same with the read order specified in in FetchShuffleBlocks
+      // because the shuffle data's return order should match the `blockIds`'s order to ensure
+      // blockId and data match.
+      if (!batchFetchEnabled) {
+        for (int j = 0; j < reduceIdArr[i].length; j++) {
+          this.blockIds[blockIdIndex++] = "shuffle_" + shuffleId + "_" + mapIds[i] + "_"
+                  + reduceIdArr[i][j];
```

Comment:
How about we put `blockId` to the `mapIdToReduceIds` first and parse it later here? So we only reorder the `blockIds` and don't need to reconstruct it.

Reply:
Well, I also think reconstruct is not well, a new `mapIdToBlockIds` which has the same key order with `mapIdToReduceIds` is needed, what's your opinion?

Reply:
We don't need an extra `mapIdToBlockIds`. We should just replace `mapIdToReduceIds: HashMap<Long, ArrayList<Integer>>` to `mapIdToBlockIds: HashMap<Long, ArrayList<String>>`.

Reply:
I see. It seems to be feasible.

Reply:
No...we can move the parse step here.

Reply:
It will parse mapId at first, then fill the `mapIdToBlockIds`; 
second, parse the reduceId and fill `ShuffleFetchBlocks` `reduceIdArr`

Reply:
OK I see. You mean we need to retrieve mapId twice.

Reply:
We should splits blockId to fill `mapIdToBlockIds` and then split blockId again to fill the `reduceIdArr` in `ShuffleFetchBlocks`

Reply:
Then, how about we carry the raw `blockId` with the splited block parts? For example, using a String array and always put the raw `blockId` at index 0 to be compatible between batchfetch and non-batchfetch. 

BTW, I saw your latest updates, and I dont' think the `LinkedHashMap` helps anything in the current way(option1), while it does help in option2.

Reply:
We can consider the proposal from @cloud-fan . I think it maybe more flexible.
(`LinkedHashMap` is used to ensure `mapIdToBlockIds.keySet()` and `mapIdToReduceIds.keySet()` keep same key order. I will use one map instead of two, so ignore this, haha)

Reply:
I think my way is similar to @cloud-fan's in general, where both two ways suggest using one map by storing the raw `blockId` together with splited block id parts(the difference is whether it's parsed).


Besides, my way avoid introducing the new class `BlocksInfo` by reusing the `blockIdParts` which is more efficient and simpler. And I don't see the flexible requirements for this function now,  we can add `BlocksInfo` whenever we needs it.

@cloud-fan WDYT?

Reply:
Do you mean replace `mapIdToReduceIds` from `HashMap<Long, ArrayList<Integer>>` to `HashMap<Long, Pair<Integer, String>>`, add extral blockId in a pair?

Reply:
@seayoun @cloud-fan  I mean: https://github.com/Ngone51/spark/pull/7/files

Reply:
I understand your design and it's a smart design, I think it maybe a little complex for reading; single class may be more easier to read.
What's your options?

Reply:
I think our original motivation was to avoid the reconstruction of `blockId`, which avoids the intermediate object as well. And now we have avoided the reconstruction but introduced another new intermediate object - `BlocksInfo`. Does it become more efficient or more readable?

As for you think https://github.com/Ngone51/spark/pull/7/files is less readable...I personally don't really think it's hard to read..perhaps adding more comments should help.